server/src/test/java/org/apache/iotdb/db/conf/adapter/ActiveTimeSeriesCounterTest.java

```diff
@@ -0,0 +1,122 @@
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+/**
+ * ActiveTimeSeriesCounter Tester.
+ */
+public class ActiveTimeSeriesCounterTest {
+
+  private static final String TEST_SG_PREFIX = "root.sg_";
+  private static int testStorageGroupNum = 10;
+  private static String[] storageGroups = new String[testStorageGroupNum];
+  private static int[] sensorNum = new int[testStorageGroupNum];
+  private static double totalSeriesNum = 0;
+
+  static {
+    for (int i = 0; i < testStorageGroupNum; i++) {
+      storageGroups[i] = TEST_SG_PREFIX + i;
+      sensorNum[i] = i + 1;
+      totalSeriesNum += sensorNum[i];
+    }
+  }
+
+  @Before
+  public void before() throws Exception {
+    ActiveTimeSeriesCounter.clear();
+    for (String storageGroup : storageGroups) {
+      ActiveTimeSeriesCounter.getInstance().init(storageGroup);
+    }
+  }
+
+  @After
+  public void after() throws Exception {
+    for (String storageGroup : storageGroups) {
+      ActiveTimeSeriesCounter.getInstance().delete(storageGroup);
+    }
+  }
+
+  /**
+   * Method: init(String storageGroup)
+   */
+  @Test
+  public void testInit() throws Exception {
+    for (int i = 0; i < testStorageGroupNum; i++) {
+      assertEquals(0D, ActiveTimeSeriesCounter.getInstance().getActiveRatio(storageGroups[i]), 0.0);
+    }
+  }
+
+  /**
+   * Method: updateActiveRatio(String storageGroup)
+   */
+  @Test
+  public void testUpdateActiveRatio() throws Exception {
+    ExecutorService service = Executors.newFixedThreadPool(storageGroups.length);
+    CountDownLatch finished = new CountDownLatch(storageGroups.length);
+    for (int i = 0; i < storageGroups.length; i++) {
+      service.submit(new OfferThreads(storageGroups[i], sensorNum[i], finished));
+    }
+    finished.await();
+    for (String storageGroup : storageGroups) {
+      ActiveTimeSeriesCounter.getInstance().updateActiveRatio(storageGroup);
+      double sum = 0;
+      for (String s : storageGroups) {
+        sum += ActiveTimeSeriesCounter.getInstance().getActiveRatio(s);
+      }
+      assertEquals(1.0, sum, 0.001);
+    }
+    for (int i = 0; i < storageGroups.length; i++) {
+      double r = ActiveTimeSeriesCounter.getInstance().getActiveRatio(storageGroups[i]);
+      assertEquals(sensorNum[i] / totalSeriesNum, r, 0.001);
+    }
+  }
+
+  private static class OfferThreads implements Runnable {
+    private int sensorNum;
+    private String storageGroup;
+    private CountDownLatch finished;
+
+    private OfferThreads(String storageGroup, int sensorNum, CountDownLatch finished) {
+      this.sensorNum = sensorNum;
+      this.storageGroup = storageGroup;
+      this.finished = finished;
+    }
+
+    @Override
+    public void run() {
+      try {
+        for (int j = 0; j < sensorNum; j++) {
+          ActiveTimeSeriesCounter.getInstance().offer(storageGroup, "device_0", "sensor_" + j);
```

Comment:
How about offering some duplicated entries for soundness?

Reply:
The duplicated entries cases is included in the HyperLogLogTest class I think.

Reply:
OK.